File: Dddml.Wms.Common/Domain/InOutDocumentActionCommandHandler.cs
using Dddml.Wms.Specialization;
using Stateless;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Dddml.Wms.Domain
{
    public class InOutDocumentActionCommandHandler : IPropertyCommandHandler<DocumentAction, string>
    {
        //[ThreadStatic]
        //private static string _currentDocumentStatus;

        //private static StateMachine<string, string> _stateMachine;

        //static InOutDocumentActionCommandHandler()
        //{
        //    var tm = BuildStateMachine(() => _currentDocumentStatus, s => _currentDocumentStatus = s);
        //    _stateMachine = tm;
        //}

        private static StateMachine<string, string> BuildStateMachine(Func<String> stateAccessor, Action<String> stateMutator)
        {
            var tm = new StateMachine<string, string>(stateAccessor, stateMutator);

            tm.Configure(DocumentStatus.Initial)
                .Permit(DocumentActionName.Draft, DocumentStatus.Drafted);

            tm.Configure(DocumentStatus.Drafted)
                .Permit(DocumentActionName.Complete, DocumentStatus.Completed)
                .Permit(DocumentActionName.Void, DocumentStatus.Voided);

            tm.Configure(DocumentStatus.Completed)
                .Permit(DocumentActionName.Close, DocumentStatus.Closed)
                .Permit(DocumentActionName.Reverse, DocumentStatus.Reversed);
            return tm;
        }

        public void Execute(IPropertyCommand<DocumentAction, string> command)
        {
            var currentState = command.GetState();
            var trigger = command.Content != null ? command.Content.Name : null;

            if (command.OuterCommandType == CommandType.Create)
            {
                if (String.IsNullOrWhiteSpace(currentState))
                { currentState = DocumentStatus.Initial; }
                if (trigger == null)
                { trigger = DocumentActionName.Draft; }
            }

            //TODO 这样才是线程安全的，但是效率不高，需要改进：
            var stateMachine = BuildStateMachine(() => currentState, command.SetState);
            stateMachine.Fire(trigger);
            //_currentDocumentStatus = currentState;
            //_stateMachine.Fire(trigger);
            //command.SetState(_currentDocumentStatus);
        }

    }
}

File: Dddml.Wms.JavaServices/src/org/dddml/wms/domain/InOutDocumentActionCommandHandler.java
package org.dddml.wms.domain;

import org.dddml.wms.specialization.CommandType;
import org.dddml.wms.specialization.PropertyCommand;
import org.dddml.wms.specialization.PropertyCommandHandler;
import org.squirrelframework.foundation.fsm.StateMachineBuilder;
import org.squirrelframework.foundation.fsm.StateMachineBuilderFactory;
import org.squirrelframework.foundation.fsm.impl.AbstractStateMachine;

/**
 * Created by Yang on 2016/7/31.
 */
public class InOutDocumentActionCommandHandler implements PropertyCommandHandler<DocumentAction, String>
{

    public static class DocumentStatusStateMachine extends AbstractStateMachine<DocumentStatusStateMachine, String, String, Object>
    {

    }

    public static StateMachineBuilder<DocumentStatusStateMachine, String, String, Object> documentStatusStateMachineBuilder;

    static
    {
        documentStatusStateMachineBuilder
                = StateMachineBuilderFactory.create(DocumentStatusStateMachine.class, String.class,String.class, Object.class);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatus.INITIAL).to(DocumentStatus.DRAFTED).on(DocumentActionName.DRAFT);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatus.DRAFTED).to(DocumentStatus.COMPLETED).on(DocumentActionName.COMPLETE);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatus.DRAFTED).to(DocumentStatus.VOIDED).on(DocumentActionName.VOID);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatus.COMPLETED).to(DocumentStatus.CLOSED).on(DocumentActionName.CLOSE);
        documentStatusStateMachineBuilder.externalTransition().from(DocumentStatus.COMPLETED).to(DocumentStatus.REVERSED).on(DocumentActionName.REVERSE);

    }

    @Override
    public final void execute(PropertyCommand<DocumentAction, String> command)
    {
        String currentState = command.getStateGetter().get();
        String trigger = command.getContent() != null ? command.getContent().getName() : null;

        if (command.getOuterCommandType().equals(CommandType.CREATE)) {
            if (currentState == null || currentState.trim().equals("")) {
                currentState = DocumentStatus.INITIAL;
            }
            if (trigger == null) {
                trigger = DocumentActionName.DRAFT;
            }
        }
        DocumentStatusStateMachine stateMachine = documentStatusStateMachineBuilder.newStateMachine(currentState);
        stateMachine.fire(trigger);
        command.getStateSetter().accept(stateMachine.getCurrentState());
    }

}
